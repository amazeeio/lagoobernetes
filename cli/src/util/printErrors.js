// @flow

import { red } from 'chalk';
import R from 'ramda';

type Cerr = typeof console.error;
type LaguError = { +message: string, +stack?: string };

export function printErrors(
  cerr: Cerr,
  ...errors: Array<LaguError | string>
): number {
  R.compose(
    R.forEach(cerr),
    R.map(red),
    R.map(
      err =>
        typeof err === 'string'
          ? err
          : R.propOr(
            // ...otherwise, use the message property as a fallback
            R.prop('message', err),
            // Try to get the stack...
            'stack',
            err,
          ),
    ),
  )(errors);
  return 1;
}

export function printNoConfigError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      '.lagoobernetes.yml config file not found. Please create one with "lagoobernetes init".\nOnline documentation: https://github.com/amazeeio/lagoobernetes/blob/master/cli/README.md#lagoobernetes-init',
  });
}

export function printProjectConfigurationError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      'No project configured. Please create a .lagoobernetes.yml config file with "lagoobernetes init" or pass a project to this command via the --project option.\nOnline documentation: https://github.com/amazeeio/lagoobernetes/blob/master/cli/README.md#lagoobernetes-init',
  });
}

export function printGraphQLErrors(
  cerr: Cerr,
  ...errors: Array<LaguError>
): number {
  if (
    R.find(R.propEq('message', 'Unauthorized - Bearer Token Required'))(errors)
  ) {
    return printErrors(cerr, {
      message:
        'No authentication token found. Please log in first with "lagoobernetes login".\nOnline documentation: https://github.com/amazeeio/lagoobernetes/blob/master/cli/README.md#lagoobernetes-login',
    });
  }
  const errorsPluralized = R.length(errors) === 1 ? 'an error' : 'errors';
  return printErrors(
    cerr,
    `Oops! The Lagoobernetes API returned ${errorsPluralized}:`,
    ...errors,
  );
}
