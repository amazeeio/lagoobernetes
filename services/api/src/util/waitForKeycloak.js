// @flow

const logger = require('../logger');
const { getKeycloakAdminClient } = require('../clients/keycloak-admin');

/* ::

type ConnectionSettings = {
  baseUrl?: string,
  realmName?: string,
  requestConfig?: Object,
}

type UserSettings = {
  username?: string,
  password?: string,
  grantType?: string,
  clientId?: string,
};

*/

async function waitForKeycloak() {
  let keycloakReady = false;
  let keycloakAdminClient;

  do {
    try {
      keycloakAdminClient = await getKeycloakAdminClient();

      if (!(await keycloakAdminClient.realms.findOne({ realm: 'lagoobernetes' }))) {
        throw new Error('The "lagoobernetes" realm has not been created yet.');
      }

      const clients = await keycloakAdminClient.clients.find({ clientId: 'api' });
      if (!clients.length) {
        throw new Error('The "api" client has not been created yet.');
      }

      keycloakReady = true;
    } catch (err) {
      logger.debug(`Waiting for Keycloak to start... (error was ${err})`);
      await new Promise(resolve => setTimeout(resolve, 5000));
    }
  } while (!keycloakReady);

  if (!keycloakAdminClient) {
    throw new Error('Keycloak client not initialized!');
  }

  logger.debug('Connected to Keycloak');
}

module.exports = waitForKeycloak;
