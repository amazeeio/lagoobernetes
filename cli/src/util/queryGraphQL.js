// @flow

import R from 'ramda';
import { getConfig } from '../config';
import { getApiConfig } from '../config/getApiConfig';
import { fileExists, readFile } from './fs';
import request from './request';
import { printErrors } from './printErrors';

type QLQueryArgs = {
  cerr: typeof console.error,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function queryGraphQL({
  cerr,
  query,
  variables,
  headers: customHeaders = {},
  pretty = false,
}:
QLQueryArgs): Object {
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders,
  };

  if (!headers.Authorization) {
    const tokenFile = R.prop('token', getConfig());
    const tokenFileExists = await fileExists(tokenFile);

    if (tokenFileExists) {
      const tokenBuffer = await readFile(tokenFile);
      const token = tokenBuffer.toString().replace(/(\r\n|\n|\r)/gm, '');
      headers.Authorization = `Bearer ${token}`;
    }
  }

  const { hostname, port } = getApiConfig();

  const protocol = R.equals(port, 443) ? 'https:' : 'http:';

  const options = {
    hostname,
    path: '/graphql',
    port,
    protocol,
    method: 'POST',
    headers,
    body: JSON.stringify(
      {
        query,
        variables,
      },
      null,
      pretty ? 2 : 0,
    ),
    rejectUnauthorized: false,
  };

  let response;

  try {
    response = await request(options);
  } catch (err) {
    const error = R.ifElse(
      // For socket hang ups...
      R.propEq('message', 'socket hang up'),
      // ...print a nicer error message...
      R.always(
        `Could not connect to Lagoobernetes API at ${protocol}//${hostname}:${port}/graphql.`,
      ),
      // ...otherwise just return the error message
      R.prop('message'),
    )(err);
    printErrors(cerr, error);
    process.exit(1);
  }

  return response;
}
