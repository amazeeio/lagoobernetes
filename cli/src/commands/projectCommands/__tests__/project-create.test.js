// @flow

import R from 'ramda';
import { queryGraphQL } from '../../../util/queryGraphQL';

import {
  CUSTOMER,
  OPENSHIFT,
  allOptionsSpecified,
  commandOptions,
  getAllowedCustomersAndOpenshifts,
  handler,
} from '../create';

jest.mock('../../../util/queryGraphQL');
jest.mock('../../../config', () => ({
  getConfig: jest.fn(() => ({ format: 'table' })),
}));

// Flow does not know which objects are actual mocks
// this function casts given parameter to JestMockFn
const _castMockForFlow = (mockFn: any): JestMockFn<any, any> => mockFn;

// Satisfy Flow types
const options = {
  ...R.omit([CUSTOMER, OPENSHIFT], commandOptions),
  [CUSTOMER]: 0,
  [OPENSHIFT]: 0,
};

describe('allOptionsSpecified', () => {
  it('should return true when all options are specified', () => {
    const returnVal = allOptionsSpecified(options);
    expect(returnVal).toBe(true);
  });

  it('should return true when all options and more are specified', () => {
    const returnVal = allOptionsSpecified({
      ...options,
      anotherOptionKey: 'anotherOptionValue',
    });
    expect(returnVal).toBe(true);
  });

  it('should return false if not all options are specified', () => {
    const returnVal = allOptionsSpecified({
      ...R.dissoc(R.head(R.keys(commandOptions)), commandOptions),
    });
    expect(returnVal).toBe(false);
  });
});

describe('getAllowedCustomersAndOpenshifts', () => {
  it('should return all customers and all openshifts', async () => {
    _castMockForFlow(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          allCustomers: [
            {
              value: 1,
              name: 'credentialtest-customer1',
            },
            {
              value: 2,
              name: 'credentialtest-customer2',
            },
          ],
          allOpenshifts: [
            {
              value: 1,
              name: 'credentialtest-openshift',
            },
          ],
        },
      }),
    );

    const cerr = jest.fn();

    const returnVal = await getAllowedCustomersAndOpenshifts(cerr);

    expect(R.prop('allCustomers', returnVal)).toHaveLength(2);
    expect(R.prop('allOpenshifts', returnVal)).toHaveLength(1);
    expect(returnVal).toMatchSnapshot();
  });
});

describe('handler', () => {
  it('should display table after successful project creation', async () => {
    _castMockForFlow(queryGraphQL)
      .mockImplementationOnce(() =>
        Promise.resolve({
          data: {
            allCustomers: [
              {
                value: 1,
                name: 'credentialtest-customer1',
              },
              {
                value: 2,
                name: 'credentialtest-customer2',
              },
            ],
            allOpenshifts: [
              {
                value: 1,
                name: 'credentialtest-openshift',
              },
            ],
          },
        }),
      )
      .mockImplementationOnce(() =>
        Promise.resolve({
          data: {
            addProject: {
              id: 1,
              name: 'test-project',
              customer: {
                name: 'credentialtest-customer1',
              },
              gitUrl: 'ssh://git@172.17.0.1:2222/git/project1.git',
              activeSystemsDeploy: 'lagoobernetes_openshiftBuildDeploy',
              activeSystemsRemove: 'lagoobernetes_openshiftRemove',
              branches: 'true',
              pullrequests: 'true',
              productionEnvironment: 'null',
              openshift: {
                name: 'credentialtest-openshift',
              },
              created: '2018-03-05 10:26:22',
            },
          },
        }),
      );

    const clog = jest.fn();
    const cerr = jest.fn();

    const returnVal = await handler({
      clog,
      cerr,
      cwd: 'some/path',
      options: {
        format: 'table',
        token: 'token/path',
        customer: 1,
        name: 'test-project',
        gitUrl: 'ssh://git@172.17.0.1:2222/git/project1.git',
        openshift: 1,
        branches: 'true',
        pullrequests: 'true',
        productionEnvironment: 'null',
      },
    });

    expect(returnVal).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it("should display error, if GraphQL sends error messages the first time it's called", async () => {
    _castMockForFlow(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve({
        errors: [{ message: 'something something error' }],
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const returnVal = await handler({
      clog,
      cerr,
      cwd: 'some/path',
      options: {
        format: 'table',
        token: 'token/path',
      },
    });

    expect(returnVal).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it("should display error, if GraphQL sends error messages the second time it's called", async () => {
    _castMockForFlow(queryGraphQL)
      .mockImplementationOnce(() =>
        Promise.resolve({
          data: {
            allCustomers: [
              {
                value: 1,
                name: 'credentialtest-customer1',
              },
              {
                value: 2,
                name: 'credentialtest-customer2',
              },
            ],
            allOpenshifts: [
              {
                value: 1,
                name: 'credentialtest-openshift',
              },
            ],
          },
        }),
      )
      .mockImplementationOnce(() =>
        Promise.resolve({
          errors: [{ message: 'something something error 2' }],
        }),
      );

    const clog = jest.fn();
    const cerr = jest.fn();

    const returnVal = await handler({
      clog,
      cerr,
      cwd: 'some/path',
      options: {
        format: 'table',
        token: 'token/path',
        customer: 1,
        name: 'test-project',
        gitUrl: 'ssh://git@172.17.0.1:2222/git/project1.git',
        openshift: 1,
        branches: 'true',
        pullrequests: 'true',
        productionEnvironment: 'null',
      },
    });

    expect(returnVal).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should display error, if GraphQL response contains zero customers', async () => {
    _castMockForFlow(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve({
        data: { allCustomers: [], allOpenshifts: [] },
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const returnVal = await handler({
      clog,
      cerr,
      cwd: 'some/path',
      options: {
        format: 'table',
        token: 'token/path',
      },
    });

    expect(returnVal).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should display error, if GraphQL response contains zero openshifts but one customer', async () => {
    _castMockForFlow(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          allCustomers: [
            {
              value: 1,
              name: 'credentialtest-customer1',
            },
          ],
          allOpenshifts: [],
        },
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const returnVal = await handler({
      clog,
      cerr,
      cwd: 'some/path',
      options: {
        format: 'table',
        token: 'token/path',
      },
    });

    expect(returnVal).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });
});
