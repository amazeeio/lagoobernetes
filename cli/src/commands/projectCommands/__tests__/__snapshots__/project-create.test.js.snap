// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllowedCustomersAndOpenshifts should return all customers and all openshifts 1`] = `
Object {
  "allCustomers": Array [
    Object {
      "name": "credentialtest-customer1",
      "value": 1,
    },
    Object {
      "name": "credentialtest-customer2",
      "value": 2,
    },
  ],
  "allOpenshifts": Array [
    Object {
      "name": "credentialtest-openshift",
      "value": 1,
    },
  ],
  "errors": undefined,
}
`;

exports[`handler should display error, if GraphQL response contains zero customers 1`] = `
Array [
  Array [
    "[31mNo authorized customers found![39m",
  ],
]
`;

exports[`handler should display error, if GraphQL response contains zero openshifts but one customer 1`] = `
Array [
  Array [
    "[31mNo authorized openshifts found![39m",
  ],
]
`;

exports[`handler should display error, if GraphQL sends error messages the first time it's called 1`] = `
Array [
  Array [
    "[31mOops! The Lagoobernetes API returned an error:[39m",
  ],
  Array [
    "[31msomething something error[39m",
  ],
]
`;

exports[`handler should display error, if GraphQL sends error messages the second time it's called 1`] = `
Array [
  Array [
    "[31mOops! The Lagoobernetes API returned an error:[39m",
  ],
  Array [
    "[31msomething something error 2[39m",
  ],
]
`;

exports[`handler should display table after successful project creation 1`] = `
Array [
  Array [
    "[34m![39m Set \\"customer\\" option to \\"1\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"name\\" option to \\"test-project\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"gitUrl\\" option to \\"ssh://git@172.17.0.1:2222/git/project1.git\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"openshift\\" option to \\"1\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"branches\\" option to \\"true\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"pullrequests\\" option to \\"true\\" from arguments or config",
  ],
  Array [
    "[34m![39m Set \\"productionEnvironment\\" option to \\"null\\" from arguments or config",
  ],
  Array [
    "[32mProject \\"test-project\\" created successfully:[39m",
  ],
  Array [
    "[38;2;132;132;132mName[39m                   test-project
[38;2;132;132;132mCustomer[39m               credentialtest-customer1
[38;2;132;132;132mGit URL[39m                ssh://git@172.17.0.1:2222/git/project1.git
[38;2;132;132;132mActive Systems Deploy[39m  lagoobernetes_openshiftBuildDeploy
[38;2;132;132;132mActive Systems Remove[39m  lagoobernetes_openshiftRemove
[38;2;132;132;132mBranches[39m               true
[38;2;132;132;132mPull Requests[39m          true
[38;2;132;132;132mOpenshift[39m              credentialtest-openshift
[38;2;132;132;132mCreated[39m                2018-03-05 10:26:22",
  ],
]
`;
